### PR TITLE
ComWrappers breaking change

### LINKED ISSUE #35728: [Breaking change]: Change CreateObjectFlags.Unwrap to only unwrap wrappers from the target ComWrappers instance
### Description

Calling `GetOrCreateObjectFromComInstance` on a `System.Runtime.InteropServices.ComWrappers` instance with the `CreateObjectFlags.Unwrap` flag used to unwrap a `ManagedObjectWrapper` from any `ComWrappers` instance. Now, using this flag will only unwrap wrappers from the `ComWrappers` instance that `GetOrCreateObjectFromComInstance` was called on.

Currently, the Unwrap flag is the only API that reaches "across" ComWrappers instances, so its behavior is unintuitive. Additionally, we just released an API on ComWrappers called TryGetObject that will unwrap a COM object from any ComWrappers instance.

### Version

.NET 8 Preview 5

### Previous behavior

Calling `GetOrCreateObjectFromComInstance` on a `System.Runtime.InteropServices.ComWrappers` instance with the `CreateObjectFlags.Unwrap` flag would to unwrap a `ManagedObjectWrapper` from any `ComWrappers` instance.

### New behavior

Calling `GetOrCreateObjectFromComInstance` on a `System.Runtime.InteropServices.ComWrappers` instance with the `CreateObjectFlags.Unwrap` flag will unwrap a `ManagedObjectWrapper` from only the `ComWrappers` instance that `GetOrCreateObjectFromComInstance` was called on. If given a wrapper from a different `ComWrappers` instance, the `ComWrappers` instance will create a new wrapper.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

The prior behavior was unintuitive and broke the encapsulation experience where developers can define how COM interop works for their code by using their own custom `ComWrappers` instances.

### Recommended action

If the prior experience is desired, call the `ComWrappers.TryGetObject` method before calling `GetOrCreateObjectForComInstance`.

### Feature area

Interop

### Affected APIs

- `System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance` when passed the `CreateObjectFlags.Unwrap` flag.

---
[Associated WorkItem - 98887](https://dev.azure.com/msft-skilling/Content/_workitems/edit/98887)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #35728 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/8.0.md](https://github.com/dotnet/docs/blob/9f98269b52cb326be224681f30ab4ed7f7a2767f/docs/core/compatibility/8.0.md) | [Breaking changes in .NET 8](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/8.0?branch=pr-en-us-35759) |
| [docs/core/compatibility/interop/8.0/comwrappers-unwrap.md](https://github.com/dotnet/docs/blob/9f98269b52cb326be224681f30ab4ed7f7a2767f/docs/core/compatibility/interop/8.0/comwrappers-unwrap.md) | [docs/core/compatibility/interop/8.0/comwrappers-unwrap](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/interop/8.0/comwrappers-unwrap?branch=pr-en-us-35759) |

### CODE REVIEW

**Review by jkoritzinsky — DISMISSED** [review]

File: docs/core/compatibility/interop/8.0/comwrappers-unwrap.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+## Previous behavior
+
+Calling <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag unwrapped a <xref:Microsoft.Office.InfoPath.ManagedObjectWrapper> from any <xref:System.Runtime.InteropServices.ComWrappers> instance.
+
+## New behavior
+
+Calling <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag only unwraps a <xref:Microsoft.Office.InfoPath.ManagedObjectWrapper> from the <xref:System.Runtime.InteropServices.ComWrappers> instance that `GetOrCreateObjectForComInstance` was called on. If given a wrapper from a different <xref:System.Runtime.InteropServices.ComWrappers> instance, the `ComWrappers` instance creates a new wrapper.
```
Comment (jkoritzinsky) [inline docs/core/compatibility/interop/8.0/comwrappers-unwrap.md:18]:
```suggestion
Calling <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag only unwraps a managed object wrapper from the <xref:System.Runtime.InteropServices.ComWrappers> instance that `GetOrCreateObjectForComInstance` was called on. If given a wrapper from a different <xref:System.Runtime.InteropServices.ComWrappers> instance, the `ComWrappers` instance creates a new wrapper.
```

File: docs/core/compatibility/interop/8.0/comwrappers-unwrap.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+The <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap> flag was the only API that reached "across" <xref:System.Runtime.InteropServices.ComWrappers> instances, so its behavior was unintuitive. Additionally, the new <xref:System.Runtime.InteropServices.ComWrappers.TryGetObject(System.IntPtr,System.Object@)?displayProperty=nameWithType> API is available to unwrap a COM object from any <xref:System.Runtime.InteropServices.ComWrappers> instance.
+
+## Previous behavior
+
+Calling <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag unwrapped a <xref:Microsoft.Office.InfoPath.ManagedObjectWrapper> from any <xref:System.Runtime.InteropServices.ComWrappers> instance.
```
Comment (jkoritzinsky) [inline docs/core/compatibility/interop/8.0/comwrappers-unwrap.md:14]:
```suggestion
Calling <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag unwrapped a managed object wrapper from any <xref:System.Runtime.InteropServices.ComWrappers> instance.
```

File: docs/core/compatibility/interop/8.0/comwrappers-unwrap.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+---
+title: "Breaking change: CreateObjectFlags.Unwrap only unwraps on target instance"
+description: Learn about the breaking change in interop in .NET 8 where 'GetOrCreateObjectForComInstance()' with the 'CreateObjectFlags.Unwrap' flag only unwraps wrappers from the target 'ComWrappers' instance.
+ms.date: 06/12/2023
+---
+# CreateObjectFlags.Unwrap only unwraps on target instance
+
+Previously, if you called <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)?displayProperty=fullName?displayProperty=nameWithType> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag, a <xref:Microsoft.Office.InfoPath.ManagedObjectWrapper> was unwrapped from *any* <xref:System.Runtime.InteropServices.ComWrappers> instance. Now when the flag is specified, only wrappers from the <xref:System.Runtime.InteropServices.ComWrappers> instance that `GetOrCreateObjectFromComInstance` was called on are unwrapped.
```
Comment (jkoritzinsky) [inline docs/core/compatibility/interop/8.0/comwrappers-unwrap.md:8]:
```suggestion
Previously, if you called <xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateObjectForComInstance(System.IntPtr,System.Runtime.InteropServices.CreateObjectFlags)?displayProperty=fullName?displayProperty=nameWithType> on a <xref:System.Runtime.InteropServices.ComWrappers> instance with the <xref:System.Runtime.InteropServices.CreateObjectFlags.Unwrap?displayProperty=nameWithType> flag, a managed object wrapper was unwrapped from *any* <xref:System.Runtime.InteropServices.ComWrappers> instance. Now when the flag is specified, only wrappers from the <xref:System.Runtime.InteropServices.ComWrappers> instance that `GetOrCreateObjectFromComInstance` was called on are unwrapped.
```

**Review by jkoritzinsky — CHANGES_REQUESTED** [review]
"managed object wrapper" is a term here, not a reference to the public ManagedObjectWrapper type.

File: docs/core/compatibility/interop/8.0/comwrappers-unwrap.md (hunk 1 above)
Comment (gewarren) [reply docs/core/compatibility/interop/8.0/comwrappers-unwrap.md]: Good catch, thanks!

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**